File: benchmark/fusion/algprog/page58.f
Config SampleIntMin = -2;
Config SampleIntMax = 2;
Config NonLinear = true;
Inductive List = nil Unit | cons {Int, List};

tri = \op: Int -> Int. fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} ->
    let tail = (fix (
      \g: List -> List. \ys: List.
      match ys with
        nil _ -> nil unit
      | cons {h, t} -> cons {op h, g t}
      end
    )) (f t) in cons {h, tail}
  end
);

@Input w: Int;
op = \x: Int. * x w;

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

main = \xs: List. sum (tri op xs);

File: ml-benchmark/fusion/algprog/page58.ml
type l = Nil of unit | Cons of int * l

let tri op =
  let rec f xs =
    match xs with
    | Nil _ -> Nil ()
    | Cons (h, t) ->
      let tail =
        (let rec g ys =
           match ys with
           | Nil _ -> Nil ()
           | Cons (h2, t2) -> Cons (op h2, g t2)
         in g) (f t)
      in
      Cons (h, tail)
  in
  f

let w = read_int ()

let op x = x * w

let sum =
  let rec f xs =
    match xs with
    | Nil _ -> 0
    | Cons (h, t) -> h + f t
  in
  f

let program xs = sum (tri op xs)